File: lists/app/assets/javascripts/angular-app/app/models/Task.js
var Task = function(task_hash){
  this.id = task_hash.id;
  this.name = task_hash.name;
  this.description = task_hash.description;
  this.status = task_hash.status;
  this.dueDateString = task_hash.due_date_string;
  this.dueDate = new Date(this.dueDateString);
  this.listId = task_hash.list_id;
  this.tags = task_hash.tags;
  this.errors = task_hash.errors;

  this.displayDueDate = function(){
    return this.dueDate.toDateString() 
         + (this.dueDate < Date.now() ? ' (OVERDUE)' : '' )
  }

  this.tagNames = function(){
    return this.tags.map(x => x.name).join(", ");
  }
};

File: lists/app/models/task.rb
class Task < ApplicationRecord
  belongs_to :list
  has_many :task_tags
  has_many :tags, through: :task_tags

  def tags_attributes=(tag_attributes)
    tag = Tag.find_or_create_by(tag_attributes.values[0]) if tag_attributes.values[0]["name"].match(/[a-zA-Z]/)
    tags << tag if tag and not tags.include?(tag)
  end

  def due_date_string
    due_date.to_s
  end

  def serialize_with_all_tags
    {
      id: id,
      name: name,
      description: description,
      due_date_string: due_date_string,
      status: status,
      tags: tags.map(&:id),
      all_tags: Tag.all.map{|tag| {id: tag.id, name: tag.name}}
    }
  end
end
